janus_core/geom_opt.py

```diff
@@ -0,0 +1,94 @@
+
+from ase.optimize import LBFGS
+
+
+def optimize(
+    atoms: Atoms,
+    fmax: float = 0.1,
+    dyn_kwargs: Optional[dict[str, Any]] = None,
+    filter_func: Optional[callable] = DefaultFilter,
+    filter_kwargs: Optional[dict[str, Any]] = None,
+    optimizer: callable = LBFGS,
+    opt_kwargs: Optional[dict[str, Any]] = None,
+    struct_kwargs: Optional[dict[str, Any]] = None,
+    traj_kwargs: Optional[dict[str, Any]] = None,
+) -> Atoms:
+    """Optimize geometry of input structure.
+
+    Parameters
+    ----------
+    atoms : Atoms
+        Atoms object to optimize geometry for.
+    fmax : float
+        Set force convergence criteria for optimizer in units eV/Å.
+    dyn_kwargs : Optional[dict[str, Any]]
+        kwargs to pass to dyn.run. Default is None.
+    filter_func : Optional[callable]
+        Apply constraints to atoms through ASE filter function.
+        Default is `FrechetCellFilter`.
+    filter_kwargs : Optional[dict[str, Any]]
+        kwargs to pass to filter_func. Default is None.
+    optimzer : callable
+        ASE optimization function. Default is `LBFGS`.
+    opt_kwargs : Optional[dict[str, Any]]
+        kwargs to pass to optimzer. Default is None.
+    struct_kwargs : Optional[dict[str, Any]]
+        kwargs to pass to ase.io.write to save optimized structure.
+        Must include "filename" keyword. Default is None.
+    traj_kwargs : Optional[dict[str, Any]]
+        kwargs to pass to ase.io.write to save optimization trajectory.
+        Must include "filename" keyword. Default is None.
+
+    Returns
+    -------
+    atoms: Atoms
+        Structure with geometry optimized.
+    """
+    dyn_kwargs = dyn_kwargs if dyn_kwargs else {}
+    filter_kwargs = filter_kwargs if filter_kwargs else {}
+    opt_kwargs = opt_kwargs if opt_kwargs else {}
+    struct_kwargs = struct_kwargs if struct_kwargs else {}
+    traj_kwargs = traj_kwargs if traj_kwargs else {}
+
+    if struct_kwargs and "filename" not in struct_kwargs:
+        raise ValueError("'filename' must be included in struct_kwargs")
+
+    if traj_kwargs and "filename" not in traj_kwargs:
+        raise ValueError("'filename' must be included in traj_kwargs")
+
+    if traj_kwargs and "trajectory" not in opt_kwargs:
+        raise ValueError(
+            "'trajectory' must be a key in opt_kwargs to save the trajectory."
+        )
+
+    if filter_func is not None:
+        filtered_atoms = filter_func(atoms, **filter_kwargs)
+        dyn = optimizer(filtered_atoms, **opt_kwargs)
+    else:
+        dyn = optimizer(atoms, **opt_kwargs)
+
+    dyn.run(fmax=fmax, **dyn_kwargs)
+
+    # Write out optimized structure
+    if struct_kwargs:
+        write(images=atoms, **struct_kwargs)
+
+    # Reformat trajectory file from binary
+    if traj_kwargs:
+        traj = read(opt_kwargs["trajectory"], index=":")
+        write(images=traj, **traj_kwargs)
+        Path(opt_kwargs["trajectory"]).unlink()
```

Comment:
will this fail if the path does not exist?

Reply:
Yes it will throw a `FileNotFoundError`, but I think there would have to have been an error in writing out the trajectory first during optimisation, so I think we probably want it to throw an error if it somehow gets here?

Reply:
actually on second thought I will let the .traj on the hdd anyhow and the user can clean it up...

Reply:
As in just remove `Path(opt_kwargs["trajectory"]).unlink()`, or get rid of the entire reload/saving?

Reply:
just the uinlink line.